File: Fibonacci.java
public class Fibonacci
{
	public static void main(String[] args)
	{
		if (args.length == 0)
		{
			System.out.println("Usage: java Fibonacci <number> [number...]");
			System.out.println("Anything past the 92nd number will fail.");
		}
		else
		{
			for (int i = 0; i < args.length; i++)
			{
				int num = Integer.parseInt(args[i]);
				if (num < 93)
				{
					System.out.println("\n" + num + "th Fibonacci number");
					System.out.println("Fast: " + fast(num));
					System.out.println("Slow: " + slow(num));
				}
				else
				{
					System.out.println("This implmentation maxes out at the 92nd number: " + fast(92));
				}
				System.out.println();
			}
		}
	}
	
	private static long fast(int nth)
	{
		if (nth < 3)
		{
			return 1;
		}
		long two_previous = 1;
		long one_previous = 1;
		long current = 2;
		for (int i = 2; i < nth; i++)
		{
			current = two_previous + one_previous;
			two_previous = one_previous;
			one_previous = current;
		}
		return current;
	}
	
	private static long slow(int nth)
	{
		if (nth < 3)
		{
			return 1;
		}
		return slow(nth - 2) + slow(nth - 1);
	}
}

File: fibonacci.rb
class Fibonacci

	# returns the nth Fibonacci number
	# complexity: O(N)
	def self.fib_num(nth)
		return 1 if nth < 3
		two_previous, one_previous = 1, 1
		current = nil
		(nth - 2).times do
		  current = two_previous + one_previous
		  two_previous = one_previous
		  one_previous = current
	  end
	  current
	end

  # returns the nth Fibonacci number
  # complexity: O(2^N)
	def self.bad_fib_num(nth)
	  return 1 if nth < 3
	  return bad_fib_num(nth - 2) + bad_fib_num(nth - 1)
	end
	
end

if ARGV.count == 0
  puts "Usage: ruby fibonacci.rb <number> [<number>...]"
else
  ARGV.each do |n|
    num = n.to_i
    puts "\n#{num}th Fibonacci number:"
    puts "Fast: #{Fibonacci.fib_num(num)}"
    puts "Slow: #{Fibonacci.bad_fib_num(num)}"
  end
  puts
end
